### ISSUE TITLE
Cleanup/refactor auth capabilities

### ISSUE THREAD

**Issue by tigrannajaryan** [open-telemetry/opentelemetry-collector#2603]
We currently have some support for authentication in the Collector
- Receivers can individually add support for incoming requests to be authenticated using request Context. configauth provides a helper for all gRPC receivers to use this.
- We have Authenticator interface and oidcAuthenticator implementation. This is extensible, but only in the Core repo. It is not possible to implement additional Authenticators in contrib.
- Exporters can individually add support for outgoing requests to be authenticated using PerRPCAuthConfig. All GRPCClientSetting-based exporters get this option automatically.

Some initial thoughts on what we need to cover:

### Receiver Authentication
- Ability to define incoming request authenticators outside of the core repo.
- The authentication should be preferably orthogonal to the receiver choice. All gRPC or HTTP receivers should be able to use any request authentication implementation without doing anything special in the receiver.

### Exporter Authentication
- Similar requires as the receiver auth above.

### Multi-tenancy
- We need to be able to pass relevant information from the incoming request context through the pipeline. This is needed for multi-tenancy use-cases: https://github.com/open-telemetry/opentelemetry-collector/pull/2495
- End user should be able to specify what parts of the incoming request are the token that should be passed through (e.g. choose which headers to preserve). The token may be multi-part (?) (e.g. multiple headers).
- Exporter helper to perform batching/groupby the chosen authentication token.
- The auth token in the pipeline likely needs to be preserved in ResourceX structs to have the most chances to survive processing pipeline. This probably requires patching generated Protobufs to add the RequestContext as a non-Protobuf, in-memory field.

This also solves the use-case with processors: https://github.com/open-telemetry/opentelemetry-collector/issues/2401

Relates issues:

- [x] #2101 - Custom authenticator logic
- [ ] #2733 - Auth data as part of the resource
- [ ] #2734 - Propagate auth data to downstream services
- [ ] #3552 - Last breaking changes before GA

**Comment by tigrannajaryan** [comment]
@jpkrohling please let me know what you think. Let's discuss.

**Comment by jpkrohling** [comment]
> Ability to define incoming request authenticators outside of the core repo.

This is the trickiest so far. The idea with having authenticators as extensions was a good one, but the configauth isn't a component, so, it can't see the available extensions. Ideas?

> The authentication should be preferably orthogonal to the receiver choice. All gRPC or HTTP receivers should be able to use any request authentication implementation without doing anything special in the receiver.

I believe we have this today already, at least for gRPC. The authenticator package provides helper functions with interceptors for gRPC, but other transports/consumers can just use the `#Authenticate(context.Context, map[string]string)` function directly if needed:

https://github.com/open-telemetry/opentelemetry-collector/blob/0345f0d275bf6814aa95fe0ca1ce08860b24791e/config/configauth/authenticator.go#L49-L51

https://github.com/open-telemetry/opentelemetry-collector/blob/0345f0d275bf6814aa95fe0ca1ce08860b24791e/config/configauth/authenticator.go#L36-L37

The interceptor for HTTP receivers isn't done yet mostly because I wanted to start with one implementation that I would have a consumer already, just to validate the idea. Adding support to all HTTP receivers shouldn't be a problem.

> Similar requires as the receiver auth above.

This could be trickier, as exporters are using their own HTTP client, from what I remember. If we don't yet, we should provide a facility to return an "instrumented" HTTP Client, which would get the auth data injected when configured.

> We need to be able to pass relevant information from the incoming request context through the pipeline.

We do store the subject/membership in the context already:
https://github.com/open-telemetry/opentelemetry-collector/blob/0345f0d275bf6814aa95fe0ca1ce08860b24791e/config/configauth/oidc_authenticator.go#L108

https://github.com/open-telemetry/opentelemetry-collector/blob/0345f0d275bf6814aa95fe0ca1ce08860b24791e/config/configauth/oidc_authenticator.go#L114

Granted, we need a component to consume this info to see if what we have in place is: 1) consumable, 2) sufficient. 

> End user should be able to specify what parts of the incoming request are the token that should be passed through (e.g. choose which headers to preserve).

Do you have concrete use cases for that? The standard is to use the regular HTTP `Authorization` header. Are you saying that we should support non-standard scenarios? In any case, the `Attribute` property of the configauth allows users to specify a custom header where the auth data is present. It does not allow using multiple headers, though.

> Exporter helper to perform batching/groupby the chosen authentication token.

I think this shouldn't be a concern of the auth mechanism, but rather, of the grouping component.

> The auth token in the pipeline likely needs to be preserved in ResourceX structs to have the most chances to survive processing pipeline. This probably requires patching generated Protobufs to add the RequestContext as a non-Protobuf, in-memory field.

Not quite sure I get this part. Could you give me more hints?

**Comment by jpkrohling** [comment]
Can we setup an ad-hoc call to discuss this one? Everyone is invited but would be useful to have at least @tigrannajaryan and me on the call. I suggest next Wednesday, Mar 17 right before the SIG Collector call. 

cc @gramidt, @pavankrish123

**Comment by tigrannajaryan** [comment]
> Can we setup an ad-hoc call to discuss this one? Everyone is invited but would be useful to have at least @tigrannajaryan and me on the call. I suggest next Wednesday, Mar 17 right before the SIG Collector call.
> 
> cc @gramidt, @pavankrish123

@jpkrohling can you please send a Doodle for a few different days/times?

**Comment by jpkrohling** [comment]
Sure: https://doodle.com/poll/rigv7yn2trd5csb4?utm_source=poll&utm_medium=link

**Comment by jpkrohling** [comment]
@tigrannajaryan Are you available on any of the proposed slots?

**Comment by tigrannajaryan** [comment]
Submitted my availability.

**Comment by tigrannajaryan** [comment]
@jpkrohling are we having the meeting today?

**Comment by tigrannajaryan** [comment]
@jpkrohling actually, let's discuss in Collector SIG, something came up unexpectedly that I need to attend.

**Comment by jpkrohling** [comment]
Sorry, left the meeting details in Slack only, and was about to post it here when I saw your request to discuss it as part of SIG Collector later today .

**Comment by gramidt** [comment]
I currently have an overlap in meetings this morning, but this discussion is very important to me, so I'll try to end the other meeting as early as possible.

**Comment by tigrannajaryan** [comment]
@jpkrohling OK, so to recap, you will give it another try to add factory-based approach for authenticators and if you can't see a good way let me and Bogdan know to see if we can help (I'd try to define an InitializableConfig interface, and we probe for the interface and call InitConfig on it during LoadConfig). The second part is adding auth info to ResourceX so that it is passed through the pipeline. The third part is how the preserved info is injected into the outgoing HTTP requests in the exporters via a generic HTTPClient (maybe also generic GRPCclient somehow?) helper.

Does this sound right?

**Comment by tigrannajaryan** [comment]
This https://github.com/open-telemetry/opentelemetry-collector/blob/main/client/client.go is somewhat related to what we do with auth. It helps extract the incoming request's IP address and uses it to identify the sender in components like K8s tagger. The question is if this needs to be a separate concept like it is now or our new auth approach can handle this use case too?

**Comment by jpkrohling** [comment]
Alright, I created two new issues as an outcome of our discussion. We have then the following issues to work on, which should address all the points discussed here:

- [ ] #2101 - Custom authenticator logic
- [ ] #2733 - Auth data as part of the resource
- [ ] #2734 - Propagate auth data to downstream services

Please, review the labels on those, especially regarding the criticality.

**Comment by alolita** [comment]
@tigrannajaryan who owns completion of this issue? Please advise.

**Comment by tigrannajaryan** [comment]
> @tigrannajaryan who owns completion of this issue? Please advise.

I believe @jcchavezs wanted to help, not sure if that's still the plan. Otherwise @jpkrohling mentioned that they will work on it next. Can you guys tell if you are able to help?

**Comment by jpkrohling** [comment]
If @jcchavezs gets blocked or can't work on this, I'm ready to jump in.

**Comment by jpkrohling** [comment]
Yes, I'll work on it next.

**Comment by alolita** [comment]
@jpkrohling What's the status on this issue? Are there related PRs?

**Comment by jpkrohling** [comment]
I just updated the main description to include a link to the related issues. Basically, the auth on the receivers' side is done, as well as on the exporters'. The propagation of auth data down the pipeline is missing, and we are having design discussions. The missing piece is critical for multi-tenancy and for a "passthrough" auth method, so that exporters would reuse the auth data they got from receivers.

**Comment by alolita** [comment]
Hi @tigrannajaryan what else needs to be done before the tracing stability release? @jpkrohling mentioned that the auth functionality has been completed and the auth propagation logic can be done after release. Please advise.

**Comment by tigrannajaryan** [comment]
>@jpkrohling mentioned that the auth functionality has been completed and the auth propagation logic can be done after release. 

We need to be sure auth propagation will not require breaking changes to pdata or internal APIs. We do not need to implement it before the release but we need to be certain that what we plan to do after the release is non-breaking.

**Comment by alolita** [comment]
@jpkrohling can you add verification logic or tests to ensure pdata and internal APIs work. Is there a design you have in mind for the auth propagation logic?

**Comment by jpkrohling** [comment]
Absolutely. I had forgotten that we do need at least one API change. I'll go over the code and do the breaking changes that I think will be necessary for the final solution, but will try to complete the auth propagation feature before v1 anyway.

**Comment by tigrannajaryan** [comment]
@bogdandrutu mentioned that there is a https://github.com/open-telemetry/opentelemetry-collector/tree/main/client package that may also need to be refactored to use the pipeline context propagation. cc @jpkrohling

**Comment by jpkrohling** [comment]
Created #3552 with the API change I had in mind.

**Comment by bogdandrutu** [comment]
@jpkrohling this issue is hard to follow, we made some progress already. Would you be willing to close this and open a new issue with what is left to do. Please add it to the phase2 if this is blocking otherwise phase3 :)

**Comment by jpkrohling** [comment]
Created #3629
